src/Processors/QueryPlan/ExpressionStep.cpp

```diff
@@ -75,6 +75,32 @@ void ExpressionStep::updateOutputStream()
 {
     output_stream = createOutputStream(
         input_streams.front(), ExpressionTransform::transformHeader(input_streams.front().header, *actions_dag), getDataStreamTraits());
+
+    if (!getDataStreamTraits().preserves_sorting)
+        return;
+
+    FindOriginalNodeForOutputName original_node_finder(actions_dag);
+    const auto & input_sort_description = getInputStreams().front().sort_description;
+    for (size_t i = 0, s = input_sort_description.size(); i < s; ++i)
+    {
+        const auto & desc = input_sort_description[i];
+        String alias;
+        const auto & origin_column = desc.column_name;
+        for (const auto & column : output_stream->header)
```

Comment:
I think this implementation is fine. (It's `O(input_sort_description.size() * output_stream->header.size())`, is it possible to make it faster? I think we can build a reverse index `output alias -> input ` in FindOriginalNodeForOutputName ctor.)

Interesting that there could be several columns that are aliased, and we are taking the first one here.
So, looks like having a sorting as a property is not a good design. Probably, it's better to check all the possible paths (+ monotonic functions) from steps which can use the sorting.

Reply:
> I think this implementation is fine. (It's `O(input_sort_description.size() * output_stream->header.size())`, is it possible to make it faster? I think we can build a reverse index `output alias -> input ` in FindOriginalNodeForOutputName ctor.)

see https://github.com/ClickHouse/ClickHouse/pull/48427